src/core/components/Experiment/index.js

```diff
@@ -0,0 +1,158 @@
+  uiState: UIStateType,
+  WrappedComponent: Function,
+|};
+
+const initialUIState: UIStateType = {
+  abTest: undefined,
+};
+
+export function mergeExperimentProps({
+  WrappedComponent,
+  nameId,
+  AName,
+  BName,
+}: {|
+  WrappedComponent: Function,
+  nameId: string,
+  AName: string,
+  BName: string,
+|}) {
+  return function mapStateToProps(state: AppState, ownProps: Object) {
+    return {
+      ...ownProps,
+      nameId,
+      AName,
+      BName,
+      WrappedComponent,
+    };
+  };
+}
+
+export class Experiment extends React.Component<InternalProps> {
+  abTestCookie: string | void;
+
+  static defaultProps = {
+    _cookie: cookie,
+    _tracking: tracking,
+    randomizer: Math.random,
+  };
+
+  componentDidMount() {
+    const { _cookie, _tracking, AName, BName, nameId, randomizer } = this.props;
+
+    this.abTestCookie = _cookie.load(`AB_${nameId}_COOKIE`);
+
+    if (this.abTestCookie === undefined) {
+      this.abTestCookie =
+        randomizer() >= 0.5 ? `AB_${nameId}_${AName}` : `AB_${nameId}_${BName}`;
+      _cookie.save(`AB_${nameId}_COOKIE`, this.abTestCookie, {
+        path: '/',
+      });
+    }
+
+    if (this.abTestCookie) {
+      this.props.setUIState({
+        abTest: this.abTestCookie,
+      });
+
+      _tracking.sendEvent({
+        action: `${nameId} Page View`,
+        category: this.abTestCookie ? `AMO ${this.abTestCookie}` : '',
+        label: '',
+      });
+    }
+  }
+
+  trackClick = (e: SyntheticEvent<any>, url: string = '') => {
+    const { _cookie, _tracking, nameId } = this.props;
+
+    const nodeType = e.currentTarget.nodeName.toLowerCase();
+
+    if (nodeType === 'a') {
+      _tracking.sendEvent({
+        action: `${nameId} Click`,
+        category: `AMO ${_cookie.load(`AB_${nameId}_COOKIE`)}`,
+        label: url,
+      });
+    }
+  };
+
+  render() {
+    const { WrappedComponent, _cookie, nameId, uiState, ...props } = this.props;
+
+    invariant(nameId, 'nameId is required');
+
+    const exposedPropHelpers = {
+      trackClick: (...args) => this.trackClick(...args),
+    };
+
+    // When the cookie is initially set it will not be seen on client
+    // until there is a refresh so we are also checking uiState.
+    return (
+      <WrappedComponent
+        {...exposedPropHelpers}
+        {...props}
+        variant={_cookie.load(`AB_${nameId}_COOKIE`) || uiState.abTest}
+      />
+    );
+  }
+}
+
+export const extractId = (ownProps: Props) => {
+  return ownProps.nameId;
+};
+
+export function experiment({
+  _mergeExperimentProps = mergeExperimentProps,
+  nameId,
+  AName,
+  BName,
+}: {|
+  _mergeExperimentProps?: Function,
+  nameId: string,
+  AName: string,
+  BName: string,
+|}) {
+  invariant(nameId, 'nameId is required');
+  invariant(AName, 'AName is required');
+  invariant(BName, 'BName is required');
+
+  return (WrappedComponent: Function) =>
+    compose(
+      connect(
+        _mergeExperimentProps({ WrappedComponent, nameId, AName, BName }),
```

Comment:
I am wondering why you connect this component here?

Reply:
I was having some issues getting it to pass in values without some similar function. I wouldn't be surprised if there was better to do this but I saw something similar to this in another HOC and this did what I needed it to do :). The key thing was passing in the params...

I can try cleaning this up a bit more but I still connect it seems so the values will be available.

Reply:
I can give it a try but the simpler the better, and I don't see why we need to connect here. We don't use redux.

Reply:
I think you can move the whole HOC class (`Experiment`) into this function, that would give you access to all the props you want from the HOC function.

Maybe we can rename it `withExperiment`. Also, I don't think we put HOCs in the `components/` folder. If you rename the entire file with `withExperiment.js`, then you can move it to `src/core/`.

That's a really good start, once cosmetics changes have been made, we can start writing tests :)

Reply:
hmm..I tried that before but couldn't get it (note: this is a case where I have to restart the server every time it's updated/wrong :/)...but I'll recheck this one too:)

Yep - that makes sense. I'll move things around as you suggest, thanks.